### ISSUE TITLE
[Feature]: Support ClickHouse storage in Jaeger V2

### ISSUE THREAD

**Issue by haanhvu** [jaegertracing/jaeger#5058]
### Requirement

Support ClickHouse storage in [Jaeger V2](https://github.com/jaegertracing/jaeger/issues/4843)

### Problem

Not supported yet.

### Proposal

- [ ] Benchmark existing schemas and batch insert approaches
- [ ] Implement spans exporting (#4941)
- [ ] Implement spans reading
- [ ] Implement dependency store
- [ ] Add necessary storage configs and options
- [ ] Support Monitoring tab directly from ClickHouse
- [ ] Implement sampling store

**Comment by haanhvu** [comment]
@yurishkuro pls check if I miss anything

**Comment by Pushkarm029** [comment]
@haanhvu working on `Implement spans reading`.

**Comment by VaibhavMalik4187** [comment]
Hey @haanhvu I'm interested in contributing to this issue. However, I'm new to the codebase and do not have a deep understanding of the complexities of the tasks mentioned in this issue. While you're working on span export, do you think that I can start working on the `Implement dependency store` item? Looking forward to collaborating with you on these tasks. Thank!

**Comment by haanhvu** [comment]
> However, I'm new to the codebase and do not have a deep understanding of the complexities of the tasks mentioned in this issue.

These tasks are medium in complexity. You're of course welcome to work on these. But they would require more time than good first issues if you're new to Jaeger and ClickHouse.

**Comment by VaibhavMalik4187** [comment]
I'm all in, I'd appreciate it if you could share some pointers about the dependencies between each of these tasks. As a beginner, I'd like to tackle independent tasks first.

**Comment by haanhvu** [comment]
Yeah I think dependency store can be implemented independently from span store. Right @yurishkuro ?

**Comment by yurishkuro** [comment]
It can be but only as an extension of already established framework for Ch integration

**Comment by haanhvu** [comment]
> It can be but only as an extension of already established framework for Ch integration

@yurishkuro You mean #4941 still needs to be merged first?

**Comment by varshith257** [comment]
Will look into it

**Comment by varshith257** [comment]
@yurishkuro Can I take up on this issue progress? 
Or Would I prefer 
- #4868

I think this are the two major issues left to be work as part of our V2 roadmap.

**Comment by chenlujjj** [comment]
Regarding the "Implement dependency store" task, is a big data job like https://github.com/jaegertracing/spark-dependencies needed as well ?

**Comment by yurishkuro** [comment]
@chenlujjj I would say dependency store is much lower priority, but yes, ideally the spark job would be extended to support a different storage.

**Comment by zzzk1** [comment]
Working on this, I will post a design document to describe my implementation in a few days.

**Comment by adityachopra29** [comment]
Also working on it .

**Comment by zzzk1** [comment]
@yurishkuro @haanhvu I have completed a brief document describing how I plan to implement this feature. Please see [Jaeger Backend Storage ClickHouse Support](https://fast-lamprey-249.notion.site/Jaeger-Backend-Storage-ClickHouse-Support-1811f661bab78002a550e258d6e64344?pvs=4), looking for suggestions, thanks!

**Comment by yurishkuro** [comment]
@zzzk1 the main issue in your proposal is it's all formulated in terms of jaeger v1 data model, while jaeger-v1 is using OTLP data model (`ptrace`). The second issue is it looks like you're completely ignoring the results of our design exploration for the optimal schema ([Benchmark report](https://medium.com/jaegertracing/making-design-decisions-for-clickhouse-as-a-core-storage-backend-in-jaeger-62bf90a979d)) and making alternative schema choices not backed by any benchmarks.

**Comment by zzzk1** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/36791902?u=ad0191a3b19aa7b9dd9ba8669ab4e014b336c8a2&amp;v=4&amp;size=80">@zzzk1](https://github.com/zzzk1) the main issue in your proposal is it's all formulated in terms of jaeger v1 data model, while jaeger-v2 is using OTLP data model (`ptrace`). The second issue is it looks like you're completely ignoring the results of our design exploration for the optimal schema ([Benchmark report](https://medium.com/jaegertracing/making-design-decisions-for-clickhouse-as-a-core-storage-backend-in-jaeger-62bf90a979d)) and making alternative schema choices not backed by any benchmarks.

@yurishkuro Yes, I have updated the document and fixed all the issues. However, I have some questions:
1. Perhaps we should use the client: clickhouse-go first. The reason can be found in the documentation under Design -> ClickHouse Storage -> [Client](https://www.notion.so/Jaeger-Backend-Storage-ClickHouse-Support-1811f661bab78002a550e258d6e64344?pvs=4#18c1f661bab780858734c5c73f044b0e).
2. Is the origin of the `dependency` derived from `traces`? If so, do we not need to use another table for storage?
3. We introduce ClickHouse on jeager v2, should we upward compatibility with Jaeger v1?

**Comment by zzzk1** [comment]
@yurishkuro I found it confusing and rewrote it on Google Docs. Please provide your comments. [Jaeger V2: Support for ClickHouse as Storage Backend](https://docs.google.com/document/d/1xoKUxQS9Y3GgdG0aopQRdz4rs8fnY4qp7Od6b9rxK50/edit?usp=sharing)

**Comment by yurishkuro** [comment]
Good [presentation](https://fosdem.org/2025/schedule/event/fosdem-2025-5960-o11y-in-one-exploring-a-unified-telemetry-database/
) from FOSDEM'25 ([with slides](https://fosdem.org/2025/events/attachments/fosdem-2025-5960-o11y-in-one-exploring-a-unified-telemetry-database/slides/237451/O11y_in_O_K4cnpcg.pdf))

**Comment by zhengkezhou1** [comment]
I plan to split the current large PR into smaller ones because it is very difficult to review.
- [ ] Define the ClickHouse database model and deserialization model.
- [ ] Introduce the ClickHouse client and define an interface for the storage plugin to interact with ClickHouse.
- [ ] Create the corresponding factory.
- [ ] Provide the functionality to automatically create table structures.
- [ ] Implement the functionality to write traces.
- [ ] Implement the functionality to read traces.
- [ ] Complete the end-to-end (E2E) test.

**Comment by yurishkuro** [comment]
@jianliao82 please do not spam issues with unrelated comments or you will be banned from the org. If you want to test tools integration test on your own repos.
